python/orca/tutorial/xshards/images/torchvision_fineturning_exp.py

```diff
@@ -0,0 +1,213 @@
+    # train on the GPU or on the CPU, if a GPU is not available
+    device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
+
+    # our dataset has two classes only - background and person
+    num_classes = 2
+    input_path = "/Users/guoqiong/intelWork/data/PennFudanPed/"
+    data_shards = read_images(file_path=input_path + "PNGImages",
+                              target_path=input_path + "PedMasks",
+                              image_type=".png",
+                              target_type=".png")
+    collected = data_shards.collect()
+    print(collected[:5])
+
+    def get_target(image):
+        img, mask = image
+        img = img.convert("RGB")
+        mask = np.array(mask)
+        # instances are encoded as different colors
+        obj_ids = np.unique(mask)
+        # first id is the background, so remove it
+        obj_ids = obj_ids[1:]
+        # split the color-encoded mask into a set
+        # of binary masks
+        masks = mask == obj_ids[:, None, None]
+        # get bounding box coordinates for each mask
+        num_objs = len(obj_ids)
+        boxes = []
+        for i in range(num_objs):
+            pos = np.where(masks[i])
+            xmin = np.min(pos[1])
+            xmax = np.max(pos[1])
+            ymin = np.min(pos[0])
+            ymax = np.max(pos[0])
+            boxes.append([xmin, ymin, xmax, ymax])
+
+        # convert everything into a torch.Tensor
+        boxes = torch.as_tensor(boxes, dtype=torch.float32)
+        # there is only one class
+        labels = torch.ones((num_objs,), dtype=torch.int64)
+        masks = torch.as_tensor(masks, dtype=torch.uint8)
+
+        area = (boxes[:, 3] - boxes[:, 1]) * (boxes[:, 2] - boxes[:, 0])
+        # suppose all instances are not crowd
+        iscrowd = torch.zeros((num_objs,), dtype=torch.int64)
+
+        target = {}
+        target["boxes"] = boxes
+        target["labels"] = labels
+        target["masks"] = masks
+        target["area"] = area
+        target["iscrowd"] = iscrowd
+        return img, target
+
+    def get_transform(image):
+        features = image[0]
+        features = T.Resize((224, 224))(features)
+        features = T.ToTensor()(features)
+        features = features.numpy()
+        return features, image[1]
+
+    def stack_feature_labels(data):
+        def per_partition(iterator):
+            targets = {}
+            features = []
+            keys = ["boxes", "labels", "masks"]
+            for key in keys:
+                targets[key] = []
+            for it in iterator:
+                feature, target = it[0], it[1]
+                features.append(feature)
+                for key in keys:
+                    targets[key].append(target[key])
+            out = {'x': np.array(features).astype(np.float32),
+                   'y': targets}  # dictionary of list
+            return [out]
+        rdd = data.rdd.mapPartitions(lambda x: per_partition(x))
+        return SparkXShards(rdd)
```

Comment:
Actually maybe you do not need this, for this sort of logic has already been implemented in https://github.com/intel-analytics/BigDL/blob/main/python/orca/src/bigdl/orca/data/utils.py#L211
```python
def partition_get_data_label(partition_data,
                             allow_tuple=True,
                             allow_list=True,
                             has_label=True):
    """
    :param partition_data: The data partition from Spark RDD, which should be a list of records.
    :param allow_tuple: Boolean. Whether the model accepts a tuple as input. Default is True.
    :param allow_list: Boolean. Whether the model accepts a list as input. Default is True.
    :param has_label: Boolean. Whether the data partition contains labels.
    :return: Concatenated data for the data partition.
    """
    data_list = [data['x'] for data in partition_data]
    label_list = [data['y'] for data in partition_data]

    data = _convert_list_tuple(combine(data_list),
                               allow_tuple=allow_tuple, allow_list=allow_list)
    if has_label:
        label = _convert_list_tuple(combine(label_list),
                                    allow_tuple=allow_tuple, allow_list=allow_list)
    else:
        label = None

    return data, label
```
in `combine` https://github.com/intel-analytics/BigDL/blob/main/python/orca/src/bigdl/orca/data/utils.py#L178-L181:
```python
def combine(data_list):
    item = data_list[0]
    if isinstance(item, dict):
        res = {}
        for k, v in item.items():
            res[k] = np.concatenate([data[k] for data in data_list], axis=0)
```
This will combine the format that you want here as:
```
{'x': [img1, img2, ...], 'y': {'boxes': [ndarray1, ndarray2,...]}}
```

Reply:
moved to another PR